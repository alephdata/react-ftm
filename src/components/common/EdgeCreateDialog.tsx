import * as React from 'react'
import { compose } from 'redux';
import { connect, ConnectedProps } from 'react-redux';
import { defineMessages, injectIntl, WrappedComponentProps } from 'react-intl';
import { FormGroup, Intent, Button } from '@blueprintjs/core'
import { Entity, Model, Schema, Values } from '@alephdata/followthemoney'

import { IEntityContext } from 'contexts/EntityContext';
import { EdgeTypeSelect, EntitySuggest } from 'editors'
import { EdgeType } from 'types'
import { Dialog } from 'components/common'

const messages = defineMessages({
  add_link: {
    id: 'dialog.edge_create.title',
    defaultMessage: 'Add link',
  },
  source: {
    id: 'dialog.edge_create.source_label',
    defaultMessage: 'Source',
  },
  target: {
    id: 'dialog.edge_create.target_label',
    defaultMessage: 'Target',
  },
  type: {
    id: 'dialog.edge_create.type_label',
    defaultMessage: 'Type',
  },
  type_select: {
    id: 'dialog.edge_create.type_placeholder',
    defaultMessage: 'Select type',
  },
  submit: {
    id: 'dialog.edge_create.submit',
    defaultMessage: 'Create',
  },
});

interface IEdgeCreateDialogProps extends WrappedComponentProps {
  source: Entity
  target?: Entity
  isOpen: boolean,
  toggleDialog: () => any
  onSubmit: (source: Entity, target: Entity, type: EdgeType) => void
  entityContext: IEntityContext
}

interface IEdgeCreateDialogState {
  sourceQueryText: string
  targetQueryText: string
  source?: Entity
  target?: Entity
  type?: EdgeType
  isProcessing: boolean
}

class EdgeCreateDialogBase extends React.Component<IEdgeCreateDialogProps & PropsFromRedux, IEdgeCreateDialogState> {
  types: EdgeType[] = []
  state: IEdgeCreateDialogState = {
    isProcessing: false,
    sourceQueryText: '',
    targetQueryText: ''
  }

  constructor(props: IEdgeCreateDialogProps & PropsFromRedux) {
    super(props)

    this.types = EdgeType.getAll(props.model)

    this.onSelectSource = this.onSelectSource.bind(this)
    this.onSelectTarget = this.onSelectTarget.bind(this)
    this.onChangeType = this.onChangeType.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onReverse = this.onReverse.bind(this)
  }

  componentDidUpdate(prevProps: IEdgeCreateDialogProps) {
    const { isOpen, source, target } = this.props;
    if (!prevProps.isOpen && isOpen) {
      this.setState({ source, target, type: undefined });
    }
  }

  onChangeType(type: EdgeType) {
    const { source, target } = this.state
    if (source && target) {
      if (!type.match(source, target) && type.match(target, source)) {
        this.setState({ source: target, target: source })
      }
      this.setState({ type })
    }
  }

  onSelectSource(source: Entity) {
    this.setState(({ target }) => ({
      source,
      target: target?.id === source.id ? undefined : target,
      type: undefined
    }))
  }

  onSelectTarget(target: Entity) {
    this.setState(({ source }) => ({
      target,
      source: source?.id === target.id ? undefined : source,
      type: undefined
    }))
  }

  isValid() {
    const { source, target, type } = this.state
    return source && target && type && type.match(source, target)
  }

  async onSubmit(e: React.ChangeEvent<HTMLFormElement>) {
    const { onSubmit, toggleDialog } = this.props
    const { source, target, type } = this.state
    e.preventDefault()
    if (source && target && type && this.isValid()) {
      onSubmit(source, target, type);
      toggleDialog();
    }
  }

  isReversible() {
    const { source, target, type } = this.state
    return source && target && type && type.match(target, source)
  }

  onReverse() {
    const { source, target } = this.state
    if (this.isReversible()) {
      this.setState({ source: target, target: source })
    }
  }

  getTypes(): EdgeType[] {
    const { source, target } = this.state

    if (source && target) {
      return this.types.filter((et) => et.match(source, target) || et.match(target, source))
    }
    return []
  }

  getSourceLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.source).label
      }
      if (type.property) {
        return type.property.schema.label
      }
    }
  }

  getTargetLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.target).label
      }
      if (type.property) {
        return type.property.label
      }
    }
  }

  getTypeDescription(): string | undefined | null {
    const { type } = this.state
    if (type) {
      if (type.schema) {
        return type.schema.description
      }
      if (type.property) {
        return type.property.description
      }
    }
  }

  render() {
    const { entityContext, intl, isOpen, suggestionSchemata, toggleDialog } = this.props
    const { isProcessing, source, target, type, sourceQueryText, targetQueryText } = this.state
    const types = this.getTypes()

    return (
      <Dialog
        icon="new-link"
        isOpen={isOpen}
        isProcessing={isProcessing}
        title={intl.formatMessage(messages.add_link)}
        onClose={() => toggleDialog()}
        className="large"
      >
        <form onSubmit={this.onSubmit}>
          <div className="bp3-dialog-body">
            <div style={{flex: 1, display: 'flex', flexFlow: 'row'}}>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.source)} helperText={this.getSourceLabel()}>
                  <EntitySuggest
                    onSubmit={(selected: Array<Entity>) => this.onSelectSource(selected?.[0])}
                    onQueryChange={(query: string) => this.setState({ sourceQueryText: query })}
                    queryText={sourceQueryText}
                    schemata={suggestionSchemata}
                    entityContext={entityContext}
                    suggestLocalEntities={true}
                    entitySelectProps={{
                      values: source ? [source] : [],
                      allowMultiple: false
                    }}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.type)} helperText={this.getTypeDescription()}>
                  <EdgeTypeSelect
                    items={types}
                    value={type}
                    onChange={this.onChangeType}
                    placeholder={intl.formatMessage(messages.type_select)}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.target)} helperText={this.getTargetLabel()}>
                  <EntitySuggest
                    onSubmit={(selected: Array<Entity>) => this.onSelectTarget(selected?.[0])}
                    onQueryChange={(query: string) => this.setState({ targetQueryText: query })}
                    queryText={targetQueryText}
                    schemata={suggestionSchemata}
                    entityContext={entityContext}
                    suggestLocalEntities={true}
                    entitySelectProps={{
                      values: target ? [target] : [],
                      allowMultiple: false
                    }}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 0, flexShrink: 1, flexBasis: '1%'}}>
                <FormGroup label='&nbsp;'>
                  <Button
                    onClick={this.onReverse}
                    disabled={!this.isReversible()}
                    icon="exchange"
                  />
                </FormGroup>
              </div>
            </div>
          </div>
          <div className="bp3-dialog-footer">
            <div className="bp3-dialog-footer-actions">
              <Button
                intent={Intent.PRIMARY}
                disabled={!this.isValid()}
                text={intl.formatMessage(messages.submit)}
                type="submit"
              />
            </div>
          </div>
        </form>
      </Dialog>
    );
  }
}

const mapStateToProps = (state: any, ownProps: IEdgeCreateDialogProps) => {
  const { entityContext } = ownProps;
  const model = entityContext.selectModel(state);
  return ({
    model,
    suggestionSchemata: model.getSchemata()
      .filter((schema: Schema) => schema.isThing() && !schema.generated && !schema.abstract)
  });
}

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>

export const EdgeCreateDialog = connector(
  injectIntl(EdgeCreateDialogBase)
);
