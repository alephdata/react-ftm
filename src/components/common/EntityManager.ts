import {
  defaultModel,
  Entity,
  Model,
  Namespace,
  Schema,
  IEntityDatum
} from '@alephdata/followthemoney'

import { EntityChanges, EntityChangeUpdate } from 'components/common/types';
import { matchText } from 'utils';


export interface IEntityManagerProps {
  entities: Array<IEntityDatum>
  namespace?: Namespace,
  updateEntity?: (entity: Entity) => void,
  expandEntity?: (entityId: string, properties?: Array<string>, limit?: number) => Promise<any>
  resolveEntityReference?: (entityId: string) => Entity | undefined,
}

export class EntityManager {
  public readonly namespace?: Namespace
  public readonly hasExpand: boolean = false
  entities = new Map<string, Entity>()
  private overload: any = {}

  constructor(props?: IEntityManagerProps) {
    if (props) {
      const { namespace, ...rest } = props;
      this.namespace = namespace
      this.overload = rest;
      this.hasExpand = this.overload.expandEntity !== undefined;
    }

    this.getEntity = this.getEntity.bind(this);
    this.getEntities = this.getEntities.bind(this);
    this.resolveEntityReference = this.resolveEntityReference.bind(this);
  }

  getEntities(ids?: Array<string>): Entity[] {
    if (ids) {
      return ids.map(id => this.getEntity(id)).filter(e => e !== undefined) as Entity[];
    } else {
      return Array.from(this.entities.values());
    }
  }

  getThingEntities(): Entity[] {
    return this.getEntities().filter(e => !e.schema.edge);
  }

  getEntity(entityId: string): Entity | undefined {
    return this.entities.get(entityId);
  }

  hasEntity(entity: Entity): boolean {
    return this.entities.has(entity.id);
  }

  addEntities(entities: Array<Entity>) {
    entities.map(e => this.entities.set(e.id, e));
  }

  updateEntity(entity: Entity) {
    this.entities.set(entity.id, entity)

    if (this.overload?.updateEntity) {
      this.overload.updateEntity(entity);
    }

    return entity;
  }

  async expandEntity(entityId: string, properties?: Array<string>, limit?: number) {
    if (this.overload?.expandEntity) {
      const expandResults = await this.overload.expandEntity(entityId, properties, limit);
      return expandResults;
    }
  }

  resolveEntityReference(entityId: string) {
    if (this.overload?.resolveEntityReference) {
      return this.overload.resolveEntityReference(entityId);
    }
  }

  // entity changes in the reverse direction require undoing create/delete operations
  applyEntityChanges(entityChanges: EntityChanges, factor: number) {
    const { created, updated, deleted } = entityChanges;

    // created && created.forEach((entity: Entity) => factor > 0 ? this.createEntity(entity) : this.deleteEntities([entity.id]));
    updated && updated.forEach(({prev, next}: EntityChangeUpdate) => factor > 0 ? this.updateEntity(next) : this.updateEntity(prev));
    // deleted && deleted.forEach((entity: Entity) => factor > 0 ? this.deleteEntities([entity.id]) : this.createEntity(entity));
  }

  toJSON(): Array<IEntityDatum> {
    return this.getEntities().map(entity => entity.toJSON());
  }

  static fromJSON(props: any, entitiesData: Array<IEntityDatum>): EntityManager {
    const entityManager = new EntityManager(props);
    const model = new Model(defaultModel);

    const entities = entitiesData.map((entityDatum: IEntityDatum) => new Entity(model, entityDatum));

    entityManager.addEntities(entities);

    return entityManager
  }
}
