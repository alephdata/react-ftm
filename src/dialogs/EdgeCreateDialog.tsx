import * as React from 'react'
import { defineMessages, WrappedComponentProps } from 'react-intl';
import _ from 'lodash';
import { Menu, MenuItem, FormGroup, Icon, Intent, Button, Alignment, Position } from '@blueprintjs/core'
import { Select, IItemListRendererProps, IItemRendererProps } from '@blueprintjs/select';
import { Entity } from '@alephdata/followthemoney'

import { EntitySelect } from '../editors'
import { EdgeType } from '../components'
import { IGraphContext } from '../GraphContext'
import { EntityManager } from '../EntityManager'
import { Schema } from '../types';

import Dialog from './Dialog';

const messages = defineMessages({
  add_link: {
    id: 'dialog.edge_create.title',
    defaultMessage: 'Add link',
  },
  source: {
    id: 'dialog.edge_create.source_label',
    defaultMessage: 'Source',
  },
  target: {
    id: 'dialog.edge_create.target_label',
    defaultMessage: 'Target',
  },
  type: {
    id: 'dialog.edge_create.type_label',
    defaultMessage: 'Type',
  },
  type_select: {
    id: 'dialog.edge_create.type_placeholder',
    defaultMessage: 'Select type',
  },
  submit: {
    id: 'dialog.edge_create.submit',
    defaultMessage: 'Create',
  },
});

const EdgeTypeSelect = Select.ofType<EdgeType>();

interface IEdgeCreateDialogProps extends WrappedComponentProps {
  source: Entity
  target?: Entity
  isOpen: boolean,
  toggleDialog: () => any
  entityManager: EntityManager
}

interface IEdgeCreateDialogState {
  source?: Entity
  target?: Entity
  sourceSuggestions: any
  targetSuggestions: any
  type?: EdgeType
  isProcessing: boolean
}

export class EdgeCreateDialog extends React.Component<IEdgeCreateDialogProps, IEdgeCreateDialogState> {
  types: EdgeType[] = []
  state: IEdgeCreateDialogState = {
    isProcessing: false,
    sourceSuggestions: { isPending: false, results: [] },
    targetSuggestions: { isPending: false, results: [] },
  }

  constructor(props: any) {
    super(props)
    this.onSelectSource = this.onSelectSource.bind(this)
    this.onSelectTarget = this.onSelectTarget.bind(this)
    this.onChangeType = this.onChangeType.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onReverse = this.onReverse.bind(this)
    this.renderEdgeType = this.renderEdgeType.bind(this)
  }


  componentDidMount() {
    const { entityManager, source, target } = this.props
    this.setState({ source, target })
    this.types = EdgeType.getAll(entityManager.model)
  }

  componentDidUpdate(prevProps: IEdgeCreateDialogProps) {
    if (prevProps.source !== this.props.source || prevProps.target !== this.props.target) {
      const { source, target } = this.props
      this.setState({ source, target, type: undefined })
    }
  }

  onChangeType(type: EdgeType) {
    const { source, target } = this.state
    if (source && target) {
      if (!type.match(source, target) && type.match(target, source)) {
        this.setState({ source: target, target: source })
      }
      this.setState({ type })
    }
  }

  onSelectSource(source: Entity) {
    this.setState({ source, type: undefined })
  }

  onSelectTarget(target: Entity) {
    this.setState({ target, type: undefined })
  }

  isValid() {
    const { source, target, type } = this.state
    return source && target && type && type.match(source, target)
  }

  async onSubmit(e: React.ChangeEvent<HTMLFormElement>) {
    const { onSubmit } = this.props
    const { source, target, type } = this.state
    e.preventDefault()
    if (source && target && type && this.isValid()) {
      onSubmit({ source, target, type});
    }
  }

  isReversible() {
    const { source, target, type } = this.state
    return source && target && type && type.match(target, source)
  }

  onReverse() {
    const { source, target, type } = this.state
    if (this.isReversible()) {
      this.setState({ source: target, target: source })
    }
  }

  getTypes(): EdgeType[] {
    const { source, target } = this.state

    if (source && target) {
      return this.types.filter((et) => et.match(source, target) || et.match(target, source))
    }
    return []
  }

  getSourceLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.source).label
      }
      if (type.property) {
        return type.property.schema.label
      }
    }
  }

  getTargetLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.target).label
      }
      if (type.property) {
        return type.property.label
      }
    }
  }

  getTypeDescription(): string | undefined | null {
    const { type } = this.state
    if (type) {
      if (type.schema) {
        return type.schema.description
      }
      if (type.property) {
        return type.property.description
      }
    }
  }

  getEdgeTypeIcon(type?: EdgeType) {
    if (type?.schema) {
      return <Schema.Icon schema={type.schema} />
    } else {
      return <Icon icon="link" />
    }
  }

  renderEdgeTypeList(props: IItemListRendererProps<EdgeType>) {
    const { items, itemsParentRef, renderItem } = props;
    const [propertyEdgeTypes, entityEdgeTypes] = _.partition(
      items,
      (et: EdgeType) => et.isPropertyEdgeType()
    );
    return (
      <Menu ulRef={itemsParentRef}>
        {entityEdgeTypes.map(renderItem)}
        <Menu.Divider />
        {propertyEdgeTypes.map(renderItem)}
      </Menu>
    );
  }

  renderEdgeType(type: EdgeType, { handleClick, modifiers }:IItemRendererProps) {
    return <MenuItem
      active={modifiers.active}
      key={type.key}
      text={type.label}
      icon={this.getEdgeTypeIcon(type)}
      onClick={handleClick}
    />
  }

  async fetchEntitySuggestions(query: string, which?: string) {
    const stateKey = `${which}Suggestions`;

    console.log('fetching', query, which);

    const { entityManager } = this.props;
    this.setState({ [stateKey]: { isProcessing: true, results: [] } });
    const results = await entityManager.getEntitySuggestions(query, ['Thing']);
    console.log('results', which, results);
    this.setState({ [stateKey]: { isProcessing: false, results } });
  }

  render() {
    const { intl, isOpen, toggleDialog } = this.props
    const { isProcessing, source, target, type, sourceSuggestions, targetSuggestions } = this.state
    const types = this.getTypes()

    return (
      <Dialog
        icon="new-link"
        isOpen={isOpen}
        isProcessing={isProcessing}
        title={intl.formatMessage(messages.add_link)}
        onClose={() => toggleDialog()}
        className="large"
      >
        <form onSubmit={this.onSubmit}>
          <div className="bp3-dialog-body">
            <div style={{flex: 1, display: 'flex', flexFlow: 'row'}}>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.source)} helperText={this.getSourceLabel()}>
                  <EntitySelect
                    onSubmit={(selected: Array<Entity>) => this.onSelectSource(selected?.[0])}
                    onChange={(selected: Array<Entity>) => this.onSelectSource(selected?.[0])}
                    values={source ? [source] : []}
                    allowMultiple={false}
                    isFetching={sourceSuggestions.isPending}
                    entitySuggestions={sourceSuggestions.results}
                    onQueryChange={(query) => this.fetchEntitySuggestions(query, 'source')}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.type)} helperText={this.getTypeDescription()}>
                  <EdgeTypeSelect
                    popoverProps={{
                      position: Position.BOTTOM_LEFT,
                      minimal: true,
                      targetProps: {style: {width: '100%'}}
                    }}
                    filterable={false}
                    items={types}
                    itemListRenderer={this.renderEdgeTypeList}
                    itemRenderer={this.renderEdgeType}
                    onItemSelect={this.onChangeType}
                  >
                    <Button fill
                      disabled={!types.length}
                      text={type ? type.label : intl.formatMessage(messages.type_select)}
                      alignText={Alignment.LEFT}
                      icon={this.getEdgeTypeIcon(type)}
                      rightIcon='double-caret-vertical'
                    />
                  </EdgeTypeSelect>
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.target)} helperText={this.getTargetLabel()}>
                  <EntitySelect
                    onSubmit={(selected: Array<Entity>) => this.onSelectTarget(selected?.[0])}
                    onChange={(selected: Array<Entity>) => this.onSelectTarget(selected?.[0])}
                    values={target ? [target] : []}
                    allowMultiple={false}
                    isFetching={targetSuggestions.isPending}
                    entitySuggestions={targetSuggestions.results}
                    onQueryChange={(query) => this.fetchEntitySuggestions(query, 'target')}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 0, flexShrink: 1, flexBasis: '1%'}}>
                <FormGroup label='&nbsp;'>
                  <Button
                    onClick={this.onReverse}
                    disabled={!this.isReversible()}
                    icon="exchange"
                  />
                </FormGroup>
              </div>
            </div>
          </div>
          <div className="bp3-dialog-footer">
            <div className="bp3-dialog-footer-actions">
              <Button
                intent={Intent.PRIMARY}
                disabled={!this.isValid()}
                text={intl.formatMessage(messages.submit)}
                type="submit"
              />
            </div>
          </div>
        </form>
      </Dialog>
    );
  }
}
