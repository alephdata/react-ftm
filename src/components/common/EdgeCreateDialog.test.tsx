import React from 'react';
import { EdgeCreateDialog } from './EdgeCreateDialog';
import { intl } from '../../setupTests';
import { EntityManager } from './EntityManager';
import { Model, Entity } from '@alephdata/followthemoney';
import { render } from '@testing-library/react';

const model = new Model({
  schemata: {
    thing: {
      label: 'Thing',
      plural: 'Things',
      extends: [],
      schemata: [],
      properties: {},
    },
  },
  types: {},
});

const entityManager = new EntityManager({
  model,
  entities: [],
});

const source = new Entity(model, {
  schema: 'thing',
  id: 'source',
});

describe('<EdgeCreateDialog />', () => {
  it('renders correctly', () => {
    render(
      <EdgeCreateDialog
        entityManager={entityManager}
        source={source}
        intl={intl}
        isOpen={true}
        toggleDialog={() => {
          /* no-op */
        }}
        onSubmit={() => {
          /* no-op */
        }}
        fetchEntitySuggestions={() => Promise.resolve([])}
      />
    );

    expect(document.body).toMatchSnapshot();
  });
});
