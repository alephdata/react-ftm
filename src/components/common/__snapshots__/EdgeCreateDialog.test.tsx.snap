// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EdgeCreateDialog /> renders correctly 1`] = `
<body
  class="bp3-overlay-open"
>
  <div />
  <div />
  <div
    class="bp3-portal dialog-portal-container"
  >
    <div
      aria-live="polite"
      class="bp3-overlay bp3-overlay-open bp3-overlay-scroll-container"
    >
      <div
        class="bp3-overlay-start-focus-trap bp3-overlay-appear bp3-overlay-appear-active"
        tabindex="0"
      />
      <div
        class="bp3-overlay-backdrop bp3-overlay-appear bp3-overlay-appear-active"
      />
      <div
        class="bp3-dialog-container bp3-overlay-content bp3-overlay-appear bp3-overlay-appear-active"
      >
        <div
          aria-labelledby="title-bp-dialog-0"
          class="bp3-dialog Dialog large"
          role="dialog"
        >
          <div
            class="bp3-dialog-header"
          >
            <span
              aria-hidden="true"
              class="bp3-icon bp3-icon-new-link"
              icon="new-link"
              tabindex="-1"
            >
              <svg
                data-icon="new-link"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M14.5 12a2.5 2.5 0 00-2.45 2h-7.1a2.5 2.5 0 100 1h7.1a2.5 2.5 0 102.45-3zM19 5h-2V3c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1V7h2c.55 0 1-.45 1-1s-.45-1-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <h4
              class="bp3-heading"
              id="title-bp-dialog-0"
            >
              Add link
            </h4>
            <button
              aria-label="Close"
              class="bp3-button bp3-minimal bp3-dialog-close-button"
              type="button"
            >
              <span
                aria-hidden="true"
                class="bp3-icon bp3-icon-small-cross"
                icon="small-cross"
              >
                <svg
                  data-icon="small-cross"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M11.41 10l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L8.59 10 5.3 13.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l3.29-3.3 3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="Dialog__content"
          >
            <form>
              <div
                class="bp3-dialog-body"
              >
                <div
                  style="flex: 1; display: flex; flex-flow: row;"
                >
                  <div
                    style="flex-grow: 1; flex-shrink: 1; flex-basis: auto; padding-right: 1em;"
                  >
                    <div
                      class="bp3-form-group"
                    >
                      <label
                        class="bp3-label"
                      >
                        Source
                         
                        <span
                          class="bp3-text-muted"
                        />
                      </label>
                      <div
                        class="bp3-form-content"
                      >
                        <div
                          class="bp3-popover-wrapper EntitySelect bp3-fill"
                        >
                          <div
                            aria-haspopup="true"
                            class="bp3-popover-target"
                            style="width: 100%;"
                          >
                            <div
                              class=""
                            >
                              <button
                                class="bp3-button bp3-fill bp3-align-left"
                                type="button"
                              >
                                <span
                                  class="bp3-button-text"
                                >
                                  Select an entity
                                </span>
                                <span
                                  aria-hidden="true"
                                  class="bp3-icon bp3-icon-caret-down"
                                  icon="caret-down"
                                >
                                  <svg
                                    data-icon="caret-down"
                                    height="16"
                                    viewBox="0 0 16 16"
                                    width="16"
                                  >
                                    <path
                                      d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                                      fill-rule="evenodd"
                                    />
                                  </svg>
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    style="flex-grow: 1; flex-shrink: 1; flex-basis: auto; padding-right: 1em;"
                  >
                    <div
                      class="bp3-form-group"
                    >
                      <label
                        class="bp3-label"
                      >
                        Type
                         
                        <span
                          class="bp3-text-muted"
                        />
                      </label>
                      <div
                        class="bp3-form-content"
                      >
                        <span
                          class="bp3-popover-wrapper"
                        >
                          <span
                            aria-haspopup="true"
                            class="bp3-popover-target"
                            style="width: 100%;"
                          >
                            <div
                              class=""
                            >
                              <button
                                class="bp3-button bp3-disabled bp3-fill bp3-align-left"
                                disabled=""
                                tabindex="-1"
                                type="button"
                              >
                                <span
                                  aria-hidden="true"
                                  class="bp3-icon bp3-icon-link"
                                  icon="link"
                                >
                                  <svg
                                    data-icon="link"
                                    height="16"
                                    viewBox="0 0 16 16"
                                    width="16"
                                  >
                                    <path
                                      d="M4.99 11.99c.28 0 .53-.11.71-.29l6-6a1.003 1.003 0 00-1.42-1.42l-6 6a1.003 1.003 0 00.71 1.71zm3.85-2.02L6.4 12.41l-1 1-.01-.01c-.36.36-.85.6-1.4.6-1.1 0-2-.9-2-2 0-.55.24-1.04.6-1.4l-.01-.01 1-1 2.44-2.44c-.33-.1-.67-.16-1.03-.16-1.1 0-2.09.46-2.81 1.19l-.02-.02-1 1 .02.02c-.73.72-1.19 1.71-1.19 2.81 0 2.21 1.79 4 4 4 1.1 0 2.09-.46 2.81-1.19l.02.02 1-1-.02-.02c.73-.72 1.19-1.71 1.19-2.81 0-.35-.06-.69-.15-1.02zm7.15-5.98c0-2.21-1.79-4-4-4-1.1 0-2.09.46-2.81 1.19l-.02-.02-1 1 .02.02c-.72.72-1.19 1.71-1.19 2.81 0 .36.06.69.15 1.02l2.44-2.44 1-1 .01.01c.36-.36.85-.6 1.4-.6 1.1 0 2 .9 2 2 0 .55-.24 1.04-.6 1.4l.01.01-1 1-2.43 2.45c.33.09.67.15 1.02.15 1.1 0 2.09-.46 2.81-1.19l.02.02 1-1-.02-.02a3.92 3.92 0 001.19-2.81z"
                                      fill-rule="evenodd"
                                    />
                                  </svg>
                                </span>
                                <span
                                  class="bp3-button-text"
                                >
                                  Select type
                                </span>
                                <span
                                  aria-hidden="true"
                                  class="bp3-icon bp3-icon-double-caret-vertical"
                                  icon="double-caret-vertical"
                                >
                                  <svg
                                    data-icon="double-caret-vertical"
                                    height="16"
                                    viewBox="0 0 16 16"
                                    width="16"
                                  >
                                    <path
                                      d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                                      fill-rule="evenodd"
                                    />
                                  </svg>
                                </span>
                              </button>
                            </div>
                          </span>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    style="flex-grow: 1; flex-shrink: 1; flex-basis: auto; padding-right: 1em;"
                  >
                    <div
                      class="bp3-form-group"
                    >
                      <label
                        class="bp3-label"
                      >
                        Target
                         
                        <span
                          class="bp3-text-muted"
                        />
                      </label>
                      <div
                        class="bp3-form-content"
                      >
                        <div
                          class="bp3-popover-wrapper EntitySelect bp3-fill"
                        >
                          <div
                            aria-haspopup="true"
                            class="bp3-popover-target"
                            style="width: 100%;"
                          >
                            <div
                              class=""
                            >
                              <button
                                class="bp3-button bp3-fill bp3-align-left"
                                type="button"
                              >
                                <span
                                  class="bp3-button-text"
                                >
                                  Select an entity
                                </span>
                                <span
                                  aria-hidden="true"
                                  class="bp3-icon bp3-icon-caret-down"
                                  icon="caret-down"
                                >
                                  <svg
                                    data-icon="caret-down"
                                    height="16"
                                    viewBox="0 0 16 16"
                                    width="16"
                                  >
                                    <path
                                      d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                                      fill-rule="evenodd"
                                    />
                                  </svg>
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    style="flex-grow: 0; flex-shrink: 1; flex-basis: 1%;"
                  >
                    <div
                      class="bp3-form-group"
                    >
                      <label
                        class="bp3-label"
                      >
                         
                         
                        <span
                          class="bp3-text-muted"
                        />
                      </label>
                      <div
                        class="bp3-form-content"
                      >
                        <button
                          class="bp3-button bp3-disabled"
                          disabled=""
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="bp3-icon bp3-icon-exchange"
                            icon="exchange"
                            tabindex="0"
                          >
                            <svg
                              data-icon="exchange"
                              height="16"
                              viewBox="0 0 16 16"
                              width="16"
                            >
                              <path
                                d="M1.99 5.99c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.89-2-2-2zm4.15 1.86a.495.495 0 10.7-.7L5.7 5.99h5.79c.28 0 .5-.22.5-.5s-.22-.5-.5-.5H5.7l1.15-1.15a.495.495 0 10-.7-.7l-2 2c-.1.09-.16.21-.16.35s.06.26.15.35l2 2.01zm7.85-1.86c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.89-2-2-2zM9.85 8.14a.533.533 0 00-.36-.15.495.495 0 00-.35.85l1.15 1.15h-5.8c-.28 0-.5.22-.5.5s.22.5.5.5h5.79l-1.15 1.15a.495.495 0 10.7.7l2-2c.09-.09.15-.22.15-.35s-.06-.26-.15-.35l-1.98-2z"
                                fill-rule="evenodd"
                              />
                            </svg>
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="bp3-dialog-footer"
              >
                <div
                  class="bp3-dialog-footer-actions"
                >
                  <button
                    class="bp3-button bp3-disabled bp3-intent-primary"
                    disabled=""
                    tabindex="-1"
                    type="submit"
                  >
                    <span
                      class="bp3-button-text"
                    >
                      Create
                    </span>
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        class="bp3-overlay-end-focus-trap bp3-overlay-appear bp3-overlay-appear-active"
        tabindex="0"
      />
    </div>
  </div>
</body>
`;
